File: cracking_coding_interview/chap_2/2.6_palindrome.rb
def find_length(head)
  length = 1
  cur_node = head

  while cur_node.next
    length += 1
    cur_node = cur_node.next
  end

  length
end

def linked_list_palindrome(head)
  return false if !head
  return true if !head.next

  length = find_length(head)
  count = 0
  cur_node = head
  stack = []

  ## create stack with first half of list
  while count < length / 2
    stack << cur_node.val
    cur_node = cur_node.next
    count += 1
  end

  ## skip middle node if the list is an odd length
  cur_node = cur_node.next if length.odd?

  ## compare first half of list with second half of list
  until stack.empty?
    cur_stack = stack.pop
    return false if cur_stack != cur_node.val
    cur_node = cur_node.next
  end
  true
end

File: cracking_coding_interview/chap_2/2.6_palindrome.js
const findLength = head => {
  let length = 1,
      curNode = head;

  while (curNode.next) {
    length++;
    curNode = curNode.next;
  }
  return length;
};

const linkedListPalindrom = head => {
  if (!head) return false;
  if (!head.next) return true;

  let length = findLength(head),
      count = 0,
      curNode = head;
  const stack = [];

  // create stack with first half of list
  while (count < Math.floor(length/2)) {
    stack.push(curNode.val);
    curNode = curNode.next;
    count++;
  }

  // skip middle node if the list is an odd length
  if (length % 2 !== 0) curNode = curNode.next;

  // compare first half of list with second half of list
  while (stack.length > 1) {
    const curStack = stack.pop();
    if (curStack !== curNode.val) return false;
    curNode = curNode.next;
  }
  return true;
};
